### PR TITLE
SessionManager implemented with comment saving function #527

### PR DESCRIPTION
**Pull request by helvete003**
I started to work on #527 and made a first (very basic) version that saves the comments in the session file and restores them when attaching the binary again. Before going any further with it i would like to hear from you if the code is even remotely good (my first ever contribution to any project)

### CODE REVIEW

**Comment by eteran** [comment]
I'll take a look over the weekend. Thanks for your contribution!

File: include/Types.h [hunk 1]
```diff
@@ -370,5 +370,11 @@ std::ostream&>::type operator<<(std::ostream& os, T val)
 	return os;
 }
 
+/* Comment Type */
+typedef struct {
```
Comment (10110111) [inline include/Types.h]: Why should this be a `typedef` for an unnamed struct? Why not simply `struct Comment {`?

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -153,23 +156,23 @@ set(edb_INCLUDES
 	${PROJECT_SOURCE_DIR}/include/QLongValidator.h
 	${PROJECT_SOURCE_DIR}/include/QULongValidator.h
 	${PROJECT_SOURCE_DIR}/include/HexStringValidator.h
-	${PROJECT_SOURCE_DIR}/include/Instruction.h
```
Comment (10110111) [inline src/CMakeLists.txt]: Such whitespace changes shouldn't go into the commit if they are unrelated to commit's objective: they distract the reader from the "useful" changes.

File: src/Debugger.cpp [hunk 3]
```diff
@@ -2873,7 +2870,18 @@ void Debugger::set_initial_debugger_state() {
 
 	const QString filename = session_filename();
 	if(!filename.isEmpty()) {
-		load_session(filename);
+		SessionError session_error;
+		if(!SessionManager::instance().load_session(filename, session_error)) {
+			QMessageBox::warning(
+				this,
+				tr("Error Loading Session"),
+				tr(session_error.getErrorMessage().toUtf8().constData())
```
Comment (10110111) [inline src/Debugger.cpp]: 1. Why the `.toUtf8().constData()`? `QObject:tr` already takes a `QString` as input, these conversions are superfluous, or am I missing something?
2. Qt Linguist won't find the string to give to translators, so this call to `tr` will never work (but see `QT_TR_NOOP`).

File: src/session/SessionManager.cpp [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+//------------------------------------------------------------------------------
+bool SessionManager::load_session(const QString &session_file, SessionError& session_error) {
+
+  QFile file(session_file);
+	if(file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+		QByteArray json = file.readAll();
+		QJsonParseError error;
+		auto doc = QJsonDocument::fromJson(json, &error);
+		if(error.error != QJsonParseError::NoError) {
+      session_error.err = SessionError::UnknownError;
```
Comment (10110111) [inline src/session/SessionManager.cpp]: Indentation appears broken here.

File: include/Types.h (hunk 1 above)
Comment (helvete003) [reply include/Types.h]: Damn sorry, I am used to using struct with typedef so that i don't have to write struct in front of every parameter where i am using it but i forgot that i don't have to do this with c++ anymore.

File: src/Debugger.cpp (hunk 3 above)
Comment (helvete003) [reply src/Debugger.cpp]: Sorry, this wasn't my intention

File: src/CMakeLists.txt (hunk 2 above)
Comment (helvete003) [reply src/CMakeLists.txt]: Sorry, this wasn't my intention

File: src/Debugger.cpp (hunk 3 above)
Comment (helvete003) [reply src/Debugger.cpp]: This was just a workaround to shut the compiler up. I was planing on using QT_TR_NOOP but had some trouble finding out if this would be the best way to got.

File: src/session/SessionManager.cpp (hunk 4 above)
Comment (helvete003) [reply src/session/SessionManager.cpp]: Weird, this seems to be a problem with my IDE (everything is fine there) but VIM is showing the same indentation error... Sorry

File: src/Debugger.cpp (hunk 3 above)
Comment (10110111) [reply src/Debugger.cpp]: Ah, sorry, point 1) is indeed wrong, there's no such overload.

File: src/session/SessionManager.cpp (hunk 4 above)
Comment (10110111) [reply src/session/SessionManager.cpp]: It's due to tab size. On github it's 8 chars by default.

Comment (10110111) [reply src/session/SessionManager.cpp]: This just means one should be consistent in using tabs vs spaces for indentation (in EDB the syle is to prefer tabs).

Comment (helvete003) [reply src/session/SessionManager.cpp]: Looked into the settings of my IDE and the tab size is 2 spaces. The indentation comes from my IDE trying to figure out the tab size of the moved code. Should i change the size in my IDE (found nothing in the style guide)?

Comment (10110111) [reply src/session/SessionManager.cpp]: Ideally, moved code should be moved verbatim, including any whitespace. It might be impossible to accomplish depending on the editor (although Vim should be able to do it, I've never had any problem with its yank-put functions).
As for setting up the IDE, I suppose you want to set tab policy to "never expand" and tab size for display to 4 chars (for Vim it's `set noexpandtab` and `set ts=4` respectively).

**Comment by eteran** [comment]
Things are looking not bad! So, I also reviewed the patch, and certainly it is functional, and I think that we'll merge it in because this is a long wanted feature because this is a great start.  However, I plan to make some changes afterwards. A few notes:

1. I agree with all of @10110111 's thoughts, glad you made some appropriate changes.
2. an underscore followed by a capital is reserved in c++, I think I recall one variable having that in its name. For members, we prefer trailing underscores.
3. From a design point of view, instead of **duplicating** the comment data, and having the session manager aware of what comments are, I'd prefer that we make it so any object can register itself to be asked what it would like to save/restore by the manager and it just hand it a `QVariantMap` to save/load as JSON. This is more flexible and doesn't require the session manager to "know" about comments.
4. A singleton is "OK" for now, but we can do better from an API point of view. It's not tremendously different, but in edb.cpp, the model that we've been using is globals protected by namespace and scope, which are accessed by functions in `edb::v1` namespace. This is the "official API" of edb.

I'll probably merge either tomorrow or the day after (it's getting late :-P). And make some of the "fix-ups" shortly thereafter.

Once that is done, we can start to consider how to handle things like dynamically loaded memory regions :-).

Thanks again for the PR

**Comment by eteran** [comment]
Merging now, I will make some tweaks to the code in the next couple of days. Thanks for the work!
